src/Field.tsx

```diff
@@ -65,21 +41,76 @@ export interface FieldConfig {
   innerRef?: (instance: any) => void;
 }
 
-export type FieldAttributes<T> = GenericFieldHTMLAttributes &
-  FieldConfig &
-  T & { name: string };
+export interface FieldConfig {
+  /**
+   * Field component to render. Can either be a string like 'select' or a component.
+   * @deprecated
+   */
+  component?: StringElements | React.ComponentType<LegacyFieldProps>;
+  // | React.ComponentType;
 
-export function useField<Val = any>(
-  propsOrFieldName: string | FieldAttributes<Val>
+  /**
+   * Component to render. Can either be a string e.g. 'select', 'input', or 'textarea', or a component.
+   */
+  as?: React.ComponentType<FieldProps['field']> | StringElements;
+
+  /**
+   * Render prop (works like React router's <Route render={props =>} />)
+   * @deprecated
+   */
+  render?: (props: LegacyFieldProps) => React.ReactElement;
+
+  /**
+   * Children render function <Field name>{props => ...}</Field>)
+   */
+  children?: React.ReactNode | ((props: FieldProps) => React.ReactElement);
+}
+
+type StringComponentConfig<U extends StringElements> =
+  | { component: U }
+  | { as: U };
+
+// type RenderableConfigKeys = keyof FieldConfig;
+
+type PickOne<T, K extends keyof T> = Pick<T, K> &
+  { [P in keyof Omit<T, K>]?: never };
+type PickOneConfig =
+  | PickOne<FieldConfig, 'render'>
+  | PickOne<FieldConfig, 'component' | 'children'>
+  | PickOne<FieldConfig, 'as' | 'children'>;
+
+type StrictFieldConfig<T extends PickOneConfig> = T extends {
+  render: T['render'];
+}
+  ? Pick<T, 'render'>
+  : T extends { component: T['component'] }
+  ? Pick<T, 'component'> & { children?: React.ReactNode }
+  : T extends { as: T['as'] }
+  ? Pick<T, 'as'> & { children?: React.ReactNode }
+  : T;
+type AddedProps<T extends FieldConfig> = T extends StringComponentConfig<
+  infer U
+>
+  ? GenericFieldHTMLAttributes<U>
+  : T extends { as: React.ComponentType<infer P> }
+  ? Omit<P, keyof FieldInputProps<any>>
+  : {};
+
+export type FieldAttributes<T extends PickOneConfig> = AddedProps<T> &
```

Comment:
FieldAttributes doesn't appear to result in `as={MyComponent}` having the proper types.

```jsx
interface ExtraProps {
  minimumAgeOfPurchase: number;
}

const MyComponent = (props: FieldInputProps<any> & ExtraProps) => null;

// below has Error! FieldInputProps<any> is not assignable to FieldInputProps<any> & ExtraProps
<Field name="hello" as={MyComponent} minimumAgeOfPurchase={21} />
```